### PR TITLE
Improving  Describe > Two/Three Variables > Summarise  

### PR DESCRIPTION
**Pull request by Vitalis95**
Fixes partly #4952 
@rdstern , check the progress made

### CODE REVIEW

**Comment by rdstern** [comment]
@Vitalis95 and @lilyclements this is looking very promising.  Here is the first combination I have found that gives an error.

image

Note, everything works "the other way round", namely when I swap variety and fert.  Here it is the model checkbox that doesn't work.
And this is the only combination that gives a problem.  All combinations seem to work for 2 factors and for 2 variates.

**Comment by rdstern** [comment]
@Vitalis95 and @fran2or the ANOVA options are now working well.  @lilyclements has written a very nice R function, called `data_book$anova_tables2`.  It is used with a single x in the summarise 2-variables, and with 2 xs in the 3 variable situation.  (I also found it is easy, in the script window, to add more x's if you wish.

@fran2or I think you can now look at particular brianches?  Please confirm, so you can try this dialog as well?

Then please could you also improve on Lily's function - just the output.  (It is "putting nice icing on the cake!") Here is a current example:

image

a) Delete 2 lines in the display of the ANOVA table, namely the line starting `# A tibble...`, and the line starting `<chr>`.
b) Change the pringing of NA into nothing.
c) Ideally add another argument to the function called figures.  This would allow control over the number of Significant Figures in the display of the means/model (not the ANOVA table).  If you can do this, then Vitalis to add the Significant Figures control that is already in the Describe > Tables > Summaries dialog.  

Thanks

**Comment by rdstern** [comment]
@Vitalis95 next I am looking at the frequency tables. 

image

Great, that these are already acceptable!  Note that these gives 3-way frequency tables, so here we get 2 of them, one for each variables in the multiple receiver.

image

a) Could we get rid of the summary variable column?
b) Could we add a reo with the variable names?
c) Then NA here are actually the zeros.  Could there be some flexibility in what is displayed here, or could zeros be displayed instead of NA?

**Comment by rdstern** [comment]
Finally we have this sort of layout:

image

This should give 2 2-way summary tables.  The first is the mean yields for variety by village (with villages in columns), and the second is for sizes by viallge (villages still the columns).

The current results are not that:  They are as follows:

image

These are 3 one-way summary tables.  Please change into 2 2-way summary tables.

Once done, there remains the question on what to do if there are multiple summaries.  We could change and just have a single summary there.  Instead I hope it will be easy to have more tables.  So, if there was alsl min and max in the summaries, then there would be 6 2-way tables, rather than the two described above.  

Also therefore delete the checkbox for Summaries in Rows.  But add in the checkbox to add the Margins.

**Comment by rdstern** [comment]
@Vitalis95 it is getting there.  
a) Can you change the title of the dialog to Summarise 2.3 Variables
b) In correlations make it Store Correlations
c) In correlations could the name be fert_yiekl, rather than the current cor_fert, that omits the second variable
d) In the summary table it is now correct, but the second factor needs to go across.  I hope this will also fix the order of levels, just as in the general tables dialog?
e) When there are multiple summaries could there also be more multiple 2-way tables, one for each summary as well as one for each factor in the multiole receiver.  These are all therefore just many 2-way tables.  That also makes the margins easier.

**Comment by lilyclements** [comment]
@Vitalis95 point (c) is sorted in the PR [here](https://github.com/Vitalis95/R-Instat/pull/299). If you merge this in, this should _hopefully_ fix point c (but do double check!)

**Comment by rdstern** [comment]
@Vitalis95 this is great for the summary tables - and almost there:

a) Summary tables:

image

Can you just make the tables with the single receiver factor going across?

b) Frequency tables - they are 3-way as you produce here:

image

Great, but isn't this the other way round from the general frequency tables?  So it is the third variable - here fertgrp that should go across?

c) I also found that Margins gave a bug here, and later we may want percentages.  Could you keep Margins there for now, but disable.  Otherwise you won't finish in time.

d) Correlations are nicely named now.  

e) You will soon have the new version of the ANOVA function from Francis.  

This is really close to something we could merge.  (The names are also a luxury because there will clearly be a tidying issue, but it will be great to get a version of this dialog included this week.)

**Comment by rdstern** [comment]
@Vitalis95 I am just about ready to approve.
a) One small think I notice is that the 3- way frequency tables does not (yet) have the Margins control.  It is on the summary tables option (and works when you change to frwequency tables!)  But the options should be separate.  Could you please copy the margins control to the frequency tables option.

I flag for later that we will also need percentages, for the 3-way tables,  but that will be another issue.   Also the order of levels in the tables can come later.

I hope you will get the updated ANOVA.  Worth trying his function anyway, don't delete the previous code yet, because that does work, and we may return to it, if needed.  It will be good to test as soon as we can.

**Comment by Vitalis95** [comment]
@fran2or , the updated anova_tables function  brings the following error. I have also attached the code generated by the dialog 

image

```
y_col_names_list <- "yield"
purrr::walk(.x=y_col_names_list, .f= ~ data_book$anova_tables2(data="survey",  x_col_names=c("size", "variety"), y_col_name=.x, signif.stars=FALSE, sign_level=FALSE, total=FALSE))
rm(y_col_names_list)
```

**Comment by rdstern** [comment]
@Vitalis95 and @fran2or it is great this is working now.  Here is an example:

image

a) Please could the title "ANOVA of yield ~ variety: go back above the ANOVA table?  It is now underneath.
b) I am having second thoughts about the --.  Maybe blank would be better?  Up to you.  But not the NA that is still there for the F prob
c) I thought we would have < 0.001, rather than 0 for the probability?  Here is an example from Francis:

image

d) In the Means tables of yield:  (it would be good to add the s to the table, but main point is to delete the next line "Tables of Means"

**Comment by rdstern** [comment]
@Vitalis95 That's much better, and you also have @jkmusyoka there to help with final tweaks if need be:

image

a) The title is in the right place
b) The extra line of output is gone from the means
c) The fprob is <0.001 - all good.

Now to change or at least to consider:
d) There is a stray NA there still.
e) Could there be a line break between the Total line and the line about the means.
f) Check with James - should there be those same underlines to separate the Residual and the total lines?  I'm not sure.

Now I'll look at more examples:

image

Same points as above.

Three variables:
image

Need to fix, but after the message it does it, and looks fine:  Comments as for 2 variables.

image

And the other options are working well too - including interactions.

When done those minor tweaks above can you ask Lily and James also to check - as well as me, not instead!  I think we (I mean you) are through!

**Review by rdstern — DISMISSED** [review]
@Vitalis95 I have been looking forward to approving this one for a long time.  Very well done, one and all on this one.
@N-thony over to you again, after Vitalis has made the initial changes you just asked for.

File: instat/dlgDescribeTwoVariable.vb [hunk 1]
```diff
@@ -788,6 +840,9 @@ Public Class dlgDescribeTwoVariable
                 ucrSaveTable.SetSaveType(RObjectTypeLabel.Table, strRObjectFormat:=RObjectFormat.Html)
                 ucrSaveTable.SetAssignToIfUncheckedValue("last_table")
                 ucrSaveTable.SetCheckBoxText("Store Table")
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsSummariesOperator)
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsFactorOperator)
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsCrossDfFunction)
```
[the remark this answers is not in the document]
Comment (Vitalis95) [reply instat/dlgDescribeTwoVariable.vb]: done

File: instat/dlgDescribeTwoVariable.vb [hunk 2]
```diff
@@ -802,42 +857,89 @@ Public Class dlgDescribeTwoVariable
                 ucrSaveTable.SetSaveType(RObjectTypeLabel.Table, strRObjectFormat:=RObjectFormat.Html)
                 ucrSaveTable.SetAssignToIfUncheckedValue("last_table")
                 ucrSaveTable.SetCheckBoxText("Store Cor")
-                ucrBase.clsRsyntax.SetBaseROperator(clsGroupByPipeOperator4)
-                clsGroupByPipeOperator4.SetAssignToOutputObject(strRObjectToAssignTo:="last_table",
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsSummariesOperator)
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsFactorOperator)
+                ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsCrossDfFunction)
```
[the remark this answers is not in the document]
Comment (Vitalis95) [reply instat/dlgDescribeTwoVariable.vb]: done

File: instat/dlgDescribeTwoVariable.vb [hunk 3]
```diff
@@ -802,42 +857,89 @@ Public Class dlgDescribeTwoVariable
...
                 ucrChkLevSig.Visible = True
                 ucrChkTotal.Visible = True
+                ucrChkInteraction.Visible = True
+                ucrChkTotal.Location = New Point(310, 250)
+                ucrChkLevSig.Location = New Point(397, 250)
+                ucrChkMeans.Location = New Point(310, 275)
                 If ucrChkSwapXYVar.Checked Then
+                    ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsSummariesOperator)
+                    ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsFactorOperator)
+                    ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsCrossDfFunction)
```
[the remark this answers is not in the document]
Comment (Vitalis95) [reply instat/dlgDescribeTwoVariable.vb]: done

**Comment by rdstern** [comment]
@Vitalis95 all ready to approve again, when this looked odd:

image

**Review by rdstern — APPROVED** [review]
Thanks @Vitalis95 
Back to you @N-thony

**Pull request merged**